src/cedict-service/cedict.js

```diff
@@ -0,0 +1,389 @@
+/**
+ * @module CedictData
+ */
+import CedictPermanentStorage from '@lexisCs/cedict-service/cedict-permanent-storage.js'
+
+/** A class to serve data from CEDICT */
+export default class Cedict {
+  /**
+   * @param {object} configuration - An object that describes a configuration of a CEDICT data object.
+   */
+  constructor (configuration) {
+    Cedict.checkConfiguration(configuration)
+    this._configuration = configuration
+
+    /**
+     * Whether the object is ready to serve data or not.
+     *
+     * @type {boolean}
+     */
+    this.isReady = false
+
+    this._storage = null
+
+    /**
+     * If CEDICT be stored in memory this object will hold all its data.
+     *
+     * @type {{entries: object[]|Map, meta: {}}}
+     */
+    this.cedict = {
+
+      // A dictionary's metadata
+      meta: {},
+
+      metaKey: 1,
+
+      /**
+       * If data is stored in memory `entries` will keep either
+       * an array of dictionary records (if no in memory indexes will be employed) or
+       * a map of dictionary records (if in memory indexes will be used).
+       * If dictionary data will be stored in permanents storage only `entries` will be null
+       */
+      dictionary: null,
+
+      /**
+       * If in memory indexes be used, `traditionalHeadwordsIdx`
+       * will hold a map: Map<traditionalHeadword, Array[entryIndex]>. Otherwise it will be null.
+       */
+      traditionalHeadwordsIdx: null,
+
+      /**
+       * If in memory indexes be used, `simplifiedHeadwordsIdx`
+       * will hold a map: Map<simplifiedHeadword, Array[entryIndex]>. Otherwise it will be null.
+       */
+      simplifiedHeadwordsIdx: null
+    }
+
+    /*
+    If characterForm is not specified in the request we will search for records with the
+    preferred character form specified below first. If any results for that character form will be found,
+    only those ones will be returned to the client. If no results for the preferred character form
+    are in the dictionary then we will search for records with other character forms.
+     */
+    this.preferredCharacterForm = Cedict.characterForms.TRADITIONAL
+
+    /*
+    This is a character form we will fallback into if matches for the preferred one are not found.
+     */
+    this.fallbackCharacterForm = Cedict.characterForms.SIMPLIFIED
+  }
+
+  /**
+   * Checks if the configuration supplied has all the necessary information in it.
+   * If configuration is not valid it will throw an error indicating which check failed.
+   *
+   * @param {object} configuration - A JSON like configuration object.
+   */
+  static checkConfiguration (configuration) {
+    if (!configuration.storage) throw new Error('Storage tree is missing from a configuration')
```

Comment:
wondering if we want to use JSON scheme to define the configuration file and then a schema validator to check it. Maybe for a future PR?

Reply:
I think that is a great idea because adding a check for every prop like this is tedious and will be even more so.

Reply:
ok, so next PR, right?

Reply:
Yes, I think it would make most sense in the next PR, otherwise changes will be too significant and hard to track